content/en/blog/_posts/2024-08-20-websocket-transition.md

```diff
@@ -0,0 +1,113 @@
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes v1.31, we've modified the streaming protocol that a Kubernetes
+client (such as `kubectl`) uses for these commands to the more modern
+[WebSocket streaming protocol](https://datatracker.ietf.org/doc/html/rfc6455).
+The WebSocket protocol is a currently supported standardized streaming protocol
+that guarantees compatibility and interoperability with different components and
+programming languages. The WebSocket protocol is more widely supported by modern
+proxies and gateways than SPDY.
+
+## How streaming APIs work
+
+Kubernetes upgrades HTTP connections to streaming connections by adding
+specific upgrade headers to the originating HTTP request. For example, an HTTP
+upgrade request for running the `date` command on an `nginx` container within
+a cluster is similar to the following:
+
+```console
+$ kubectl exec -v=8 nginx -- date
+GET https://127.0.0.1:43251/api/v1/namespaces/default/pods/nginx/exec?command=date…
+Request Headers:
+    Connection: Upgrade
+    Upgrade: websocket
+    Sec-Websocket-Protocol: v5.channel.k8s.io
+    User-Agent: kubectl/v1.30.1 (linux/amd64) kubernetes/6911225
+```
+
+If the container runtime supports the WebSocket streaming protocol and at least
+one of the subprotocol versions (e.g. `v5.channel.k8s.io`), the server responds
+with a successful `101 Switching Protocols` status, along with the negotiated
+subprotocol version:
+
+```console
+Response Status: 101 Switching Protocols in 3 milliseconds
+Response Headers:
+    Upgrade: websocket
+    Connection: Upgrade
+    Sec-Websocket-Accept: j0/jHW9RpaUoGsUAv97EcKw8jFM=
+    Sec-Websocket-Protocol: v5.channel.k8s.io
+```
+
+At this point the TCP connection used for the HTTP protocol has changed to a
+streaming connection. Subsequent STDIN, STDOUT, and STDERR data (as well as TTY
+and process exit code data) for this shell interaction is then streamed over this
+upgraded connection.
+
+## How to use the new WebSocket streaming protocol
+
+If your cluster and kubectl are on version 1.29 or later, there are two
+control plane feature gates and two kubectl environment variables that
+govern the use of the WebSockets rather than SPDY. In Kubernetes 1.31,
+all of the following feature gates are in beta and are enabled by
+default:
+
+- [Feature gates](/docs/reference/command-line-tools-reference/feature-gates/)
+  - `TranslateStreamCloseWebsocketRequests`
+      - `.../exec`
+      - `.../attach`
+  - `PortForwardWebsockets`
+      - `.../port-forward`
+- kubectl feature control environment variables
+  - `KUBECTL_REMOTE_COMMAND_WEBSOCKETS`
+      - `kubectl exec`
+      - `kubectl cp`
+      - `kubectl attach`
+  - `KUBECTL_PORT_FORWARD_WEBSOCKETS`
+      - `kubectl port-forward`
+
+If you're connecting to an older cluster but can manage the feature gate
+settings, turn on `TranslateStreamCloseWebsocketRequests` (available in
```

Comment:
(nit)

Turn on X and Y -> Turn on both X and Y

Reply:
Fixed.